# Exploring Physiological Predictors of Obesity Using NHANES and Relational Data Models

## Introduction

The National Health and Nutrition Examination Survey (NHANES) provides comprehensive data on the health and nutritional status of the U.S. population. Given the ongoing obesity epidemic in the United States, analyzing this dataset offers a valuable opportunity to understand factors contributing to obesity.

While lifestyle choices play a significant role, physiological factors are also important contributors. This project aims to explore NHANES data to identify potential correlations and predictors related to obesity. Ultimately, these insights could help inform strategies to prevent the development of obesity in at-risk populations.

## Objective

This project aims to:

* Clean and preprocess the NHANES datasets from 2017–2020, focusing on demographics and biometrics.
* Create a relational database structure that organizes the cleaned data for easy querying and analysis.
* Explore potential correlations between physiological and lifestyle factors with obesity-related measures.
* Develop preliminary predictive models to identify key predictors of obesity using the cleaned data.
* Document the data cleaning process thoroughly to ensure reproducibility and transparency.

## Import Libraries & Setup

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import os
!pip install pyreadstat
import pyreadstat #since the data files are .xpt files, this library is needed to import the table

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 594.8/594.8 kB 3.5 MB/s eta 0:00:00a 0:00:01

[notice] A new release of pip is available: 23.1.2 -> 25.1.1
[notice] To update, run: pip install --upgrade pip


## Data Cleaning

Prior to initiating the data cleaning process, the required Python libraries are imported. Relevant NHANES datasets will be loaded sequentially at each step to maintain clarity and focus. This approach facilitates exploration of the data structure, variable types, and initial summaries, enabling a comprehensive understanding of the data's scope and quality.

### Demographics Data Cleaning

In [2]:
file_path = '2017-2020/2017-2020.P_DEMO.xpt'

df_demo, meta = pyreadstat.read_xport(file_path)

In [3]:
df_demo.head(10)

SEQN  SDDSRVYR  RIDSTATR  RIAGENDR  RIDAGEYR  RIDAGEMN  RIDRETH1  \
0  109263.0      66.0       2.0       1.0       2.0       NaN       5.0   
1  109264.0      66.0       2.0       2.0      13.0       NaN       1.0   
2  109265.0      66.0       2.0       1.0       2.0       NaN       3.0   
3  109266.0      66.0       2.0       2.0      29.0       NaN       5.0   
4  109267.0      66.0       1.0       2.0      21.0       NaN       2.0   
5  109268.0      66.0       1.0       2.0      18.0       NaN       3.0   
6  109269.0      66.0       2.0       1.0       2.0       NaN       2.0   
7  109270.0      66.0       2.0       2.0      11.0       NaN       4.0   
8  109271.0      66.0       2.0       1.0      49.0       NaN       3.0   
9  109272.0      66.0       2.0       1.0       0.0       3.0       1.0   

   RIDRETH3  RIDEXMON  DMDBORN4  ...  FIAINTRP  MIALANG  MIAPROXY  MIAINTRP  \
0       6.0       2.0       1.0  ...       2.0      NaN       NaN       NaN   
1       1.0       2.0       1.0  ...       2.0      1.0       2.0       2.0   
2       3.0       2.0       1.0  ...       2.0      NaN       NaN       NaN   
3       6.0       2.0       2.0  ...       2.0      1.0       2.0       2.0   
4       2.0       NaN       2.0  ...       2.0      NaN       NaN       NaN   
5       3.0       NaN       1.0  ...       2.0      NaN       NaN       NaN   
6       2.0       1.0       1.0  ...       2.0      NaN       NaN       NaN   
7       4.0       1.0       1.0  ...       2.0      1.0       2.0       2.0   
8       3.0       2.0       1.0  ...       2.0      1.0       2.0       2.0   
9       1.0       2.0       1.0  ...       2.0      NaN       NaN       NaN   

   AIALANGA      WTINTPRP      WTMECPRP  SDMVPSU  SDMVSTRA  INDFMPIR  
0       NaN   7891.762435   8951.815567      3.0     156.0      4.66  
1       1.0  11689.747264  12271.157043      1.0     155.0      0.83  
2       NaN  16273.825939  16658.764203      1.0     157.0      3.06  
3       1.0   7825.646112   8154.968193      2.0     168.0      5.00  
4       NaN  26379.991724      0.000000      1.0     156.0      5.00  
5       NaN  19639.221008      0.000000      1.0     155.0      1.66  
6       NaN   5906.250521   6848.271782      2.0     152.0      0.96  
7       NaN   4613.057696   4886.930378      1.0     150.0      1.88  
8       1.0   8481.589837   8658.732873      1.0     167.0       NaN  
9       NaN   7037.380216   7872.776233      1.0     155.0      0.73  

[10 rows x 29 columns]

Many of the original column names in the dataset are not immediately intuitive. To improve readability and facilitate analysis, we will first list all column names and then rename them with more straightforward, descriptive labels. The original column descriptions can be referenced here: NHANES Demographics Codebook.

Additionally, based on the project’s scope, several columns that are not relevant to our analysis will be dropped to streamline the dataset.

In [4]:
df_demo.columns

Index(['SEQN', 'SDDSRVYR', 'RIDSTATR', 'RIAGENDR', 'RIDAGEYR', 'RIDAGEMN',
       'RIDRETH1', 'RIDRETH3', 'RIDEXMON', 'DMDBORN4', 'DMDYRUSZ', 'DMDEDUC2',
       'DMDMARTZ', 'RIDEXPRG', 'SIALANG', 'SIAPROXY', 'SIAINTRP', 'FIALANG',
       'FIAPROXY', 'FIAINTRP', 'MIALANG', 'MIAPROXY', 'MIAINTRP', 'AIALANGA',
       'WTINTPRP', 'WTMECPRP', 'SDMVPSU', 'SDMVSTRA', 'INDFMPIR'],
      dtype='object')

In [5]:
df_demo.drop(['SDDSRVYR','RIDAGEMN','RIDRETH1','RIDEXMON','DMDBORN4','DMDYRUSZ','RIDEXPRG','SIALANG', 
         'SIAPROXY', 'SIAINTRP','FIALANG','FIAPROXY', 'FIAINTRP', 'MIALANG', 'MIAPROXY', 
         'MIAINTRP', 'AIALANGA','WTINTPRP', 'WTMECPRP', 'SDMVPSU', 'SDMVSTRA'],
        axis=1,inplace=True, errors='ignore')

In [6]:
df_demo.columns

Index(['SEQN', 'RIDSTATR', 'RIAGENDR', 'RIDAGEYR', 'RIDRETH3', 'DMDEDUC2',
       'DMDMARTZ', 'INDFMPIR'],
      dtype='object')

In [7]:
df_demo = df_demo.rename(columns={
    'SEQN':'participant_id',
    'RIAGENDR': 'gender',
    'RIDAGEYR' : 'age_year',
    'RIDRETH3' : 'race',
    'DMDEDUC2' : 'education_level', 
    'DMDMARTZ' : 'marital_status',
    'INDFMPIR' : 'family_income_poverty',
    'RIDSTATR' : 'interview_exam_status'
})

In [8]:
df_demo.head(10)

participant_id  interview_exam_status  gender  age_year  race  \
0        109263.0                    2.0     1.0       2.0   6.0   
1        109264.0                    2.0     2.0      13.0   1.0   
2        109265.0                    2.0     1.0       2.0   3.0   
3        109266.0                    2.0     2.0      29.0   6.0   
4        109267.0                    1.0     2.0      21.0   2.0   
5        109268.0                    1.0     2.0      18.0   3.0   
6        109269.0                    2.0     1.0       2.0   2.0   
7        109270.0                    2.0     2.0      11.0   4.0   
8        109271.0                    2.0     1.0      49.0   3.0   
9        109272.0                    2.0     1.0       0.0   1.0   

   education_level  marital_status  family_income_poverty  
0              NaN             NaN                   4.66  
1              NaN             NaN                   0.83  
2              NaN             NaN                   3.06  
3              5.0             3.0                   5.00  
4              4.0             3.0                   5.00  
5              NaN             NaN                   1.66  
6              NaN             NaN                   0.96  
7              NaN             NaN                   1.88  
8              2.0             3.0                    NaN  
9              NaN             NaN                   0.73

In [9]:
df_demo.dtypes

participant_id           float64
interview_exam_status    float64
gender                   float64
age_year                 float64
race                     float64
education_level          float64
marital_status           float64
family_income_poverty    float64
dtype: object

In [10]:
df_demo.shape

(15560, 8)

In [11]:
df_demo.isnull().sum()

participant_id              0
interview_exam_status       0
gender                      0
age_year                    0
race                        0
education_level          6328
marital_status           6328
family_income_poverty    2201
dtype: int64

In [12]:
df_demo.describe()

participant_id  interview_exam_status        gender      age_year  \
count     15560.00000           15560.000000  15560.000000  15560.000000   
mean     117042.50000               1.919023      1.503792     33.742481   
std        4491.92943               0.272808      0.500002     25.320532   
min      109263.00000               1.000000      1.000000      0.000000   
25%      113152.75000               2.000000      1.000000     10.000000   
50%      117042.50000               2.000000      2.000000     30.000000   
75%      120932.25000               2.000000      2.000000     56.000000   
max      124822.00000               2.000000      2.000000     80.000000   

               race  education_level  marital_status  family_income_poverty  
count  15560.000000      9232.000000     9232.000000           13359.000000  
mean       3.486118         3.551993        1.708622               2.405937  
std        1.622734         1.214109        2.755878               1.634346  
min        1.000000         1.000000        1.000000               0.000000  
25%        3.000000         3.000000        1.000000               1.020000  
50%        3.000000         4.000000        1.000000               1.960000  
75%        4.000000         4.000000        2.000000               3.880000  
max        7.000000         9.000000       99.000000               5.000000

Based on the NHANES data documentation, the variable RIDSTATR (renamed as interview_exam_status) indicates whether a participant completed only the interview (1) or both the interview and physical examination (2). Since this project focuses on physical and medical examination data, participants with a value of 1 will be excluded. After filtering out these participants, the interview_exam_status column will be dropped, as it will no longer provide useful information.

In [13]:
df_demo = df_demo[df_demo['interview_exam_status'] != 1]

In [14]:
df_demo.head(10)

participant_id  interview_exam_status  gender  age_year  race  \
0         109263.0                    2.0     1.0       2.0   6.0   
1         109264.0                    2.0     2.0      13.0   1.0   
2         109265.0                    2.0     1.0       2.0   3.0   
3         109266.0                    2.0     2.0      29.0   6.0   
6         109269.0                    2.0     1.0       2.0   2.0   
7         109270.0                    2.0     2.0      11.0   4.0   
8         109271.0                    2.0     1.0      49.0   3.0   
9         109272.0                    2.0     1.0       0.0   1.0   
10        109273.0                    2.0     1.0      36.0   3.0   
11        109274.0                    2.0     1.0      68.0   7.0   

    education_level  marital_status  family_income_poverty  
0               NaN             NaN                   4.66  
1               NaN             NaN                   0.83  
2               NaN             NaN                   3.06  
3               5.0             3.0                   5.00  
6               NaN             NaN                   0.96  
7               NaN             NaN                   1.88  
8               2.0             3.0                    NaN  
9               NaN             NaN                   0.73  
10              4.0             3.0                   0.83  
11              4.0             3.0                   1.20

In [15]:
df_demo.drop(['interview_exam_status'], axis = 1, inplace = True)

In [16]:
df_demo.shape

(14300, 7)

After the initial cleaning process, the resulting dataset contains approximately 143,000 rows and 7 columns.

To maintain consistency with the original NHANES coding and to support traceability, the decision was made to retain the raw categorical values for columns containing encoded data (such as gender, race/ethnicity, and education level). These encoded values follow NHANES's standard coding conventions and will be referenced using corresponding lookup tables or detailed clarifications provided in the README.

In [17]:
df_demo.head(10)

participant_id  gender  age_year  race  education_level  marital_status  \
0         109263.0     1.0       2.0   6.0              NaN             NaN   
1         109264.0     2.0      13.0   1.0              NaN             NaN   
2         109265.0     1.0       2.0   3.0              NaN             NaN   
3         109266.0     2.0      29.0   6.0              5.0             3.0   
6         109269.0     1.0       2.0   2.0              NaN             NaN   
7         109270.0     2.0      11.0   4.0              NaN             NaN   
8         109271.0     1.0      49.0   3.0              2.0             3.0   
9         109272.0     1.0       0.0   1.0              NaN             NaN   
10        109273.0     1.0      36.0   3.0              4.0             3.0   
11        109274.0     1.0      68.0   7.0              4.0             3.0   

    family_income_poverty  
0                    4.66  
1                    0.83  
2                    3.06  
3                    5.00  
6                    0.96  
7                    1.88  
8                     NaN  
9                    0.73  
10                   0.83  
11                   1.20

At this stage, missing values (NaN) in other columns are retained to maximize the number of participants included in the dataset. This approach preserves potential correlations with other lab and physical examination variables. The cleaned dataframe has been saved as a CSV file for subsequent import into a relational database.

In [18]:
df_demo.to_csv('cleaned_demographics.csv', index=False)

The same cleaning and preprocessing steps will be applied to the other tables and datasets to ensure consistency and maintain data quality for the forthcoming analysis.

### Body Measures Data Cleaning

In [19]:
file_path = '2017-2020/2017-2020.P_BMX.xpt'

df_bmx, meta = pyreadstat.read_xport(file_path)

In [20]:
df_bmx.head(10)

SEQN  BMDSTATS  BMXWT  BMIWT  BMXRECUM  BMIRECUM  BMXHEAD  BMIHEAD  \
0  109263.0       4.0    NaN    NaN       NaN       NaN      NaN      NaN   
1  109264.0       1.0   42.2    NaN       NaN       NaN      NaN      NaN   
2  109265.0       1.0   12.0    NaN      91.6       NaN      NaN      NaN   
3  109266.0       1.0   97.1    NaN       NaN       NaN      NaN      NaN   
4  109269.0       3.0   13.6    NaN      90.9       NaN      NaN      NaN   
5  109270.0       1.0   75.3    NaN       NaN       NaN      NaN      NaN   
6  109271.0       1.0   98.8    NaN       NaN       NaN      NaN      NaN   
7  109272.0       1.0    7.1    NaN      63.6       NaN     41.3      NaN   
8  109273.0       1.0   74.3    NaN       NaN       NaN      NaN      NaN   
9  109274.0       1.0  103.7    NaN       NaN       NaN      NaN      NaN   

   BMXHT  BMIHT  ...  BMXLEG  BMILEG  BMXARML  BMIARML  BMXARMC  BMIARMC  \
0    NaN    NaN  ...     NaN     NaN      NaN      NaN      NaN      NaN   
1  154.7    NaN  ...    36.3     NaN     33.8      NaN     22.7      NaN   
2   89.3    NaN  ...     NaN     NaN     18.6      NaN     14.8      NaN   
3  160.2    NaN  ...    40.8     NaN     34.7      NaN     35.8      NaN   
4    NaN    1.0  ...     NaN     NaN      NaN      1.0      NaN      1.0   
5  156.0    NaN  ...    42.6     NaN     36.1      NaN     31.0      NaN   
6  182.3    NaN  ...    40.1     NaN     42.0      NaN     38.2      NaN   
7    NaN    NaN  ...     NaN     NaN     13.0      NaN     15.5      NaN   
8  184.2    NaN  ...    41.0     NaN     41.1      NaN     30.2      NaN   
9  185.3    NaN  ...    44.0     NaN     47.0      NaN     32.0      NaN   

   BMXWAIST  BMIWAIST  BMXHIP  BMIHIP  
0       NaN       NaN     NaN     NaN  
1      63.8       NaN    85.0     NaN  
2      41.2       NaN     NaN     NaN  
3     117.9       NaN   126.1     NaN  
4       NaN       1.0     NaN     NaN  
5      91.4       NaN     NaN     NaN  
6     120.4       NaN   108.2     NaN  
7       NaN       NaN     NaN     NaN  
8      86.8       NaN    94.5     NaN  
9     109.6       NaN   107.8     NaN  

[10 rows x 22 columns]

In [21]:
df_bmx.shape

(14300, 22)

In [22]:
df_bmx.isnull().sum()

SEQN            0
BMDSTATS        0
BMXWT         225
BMIWT       13712
BMXRECUM    12830
BMIRECUM    14257
BMXHEAD     13990
BMIHEAD     14300
BMXHT        1143
BMIHT       14129
BMXBMI       1163
BMDBMIC      9551
BMXLEG       3316
BMILEG      13812
BMXARML       810
BMIARML     13813
BMXARMC       816
BMIARMC     13807
BMXWAIST     1726
BMIWAIST    13683
BMXHIP       4438
BMIHIP      13924
dtype: int64

Looking at the document associated with this dataset, each of the column and their contents were assessed. There are participants without body measurement data, which will be excluded. 

In [23]:
df_bmx = df_bmx[df_bmx['BMDSTATS'] != 4]

In [24]:
df_bmx.shape

(14107, 22)

The head circumference is measured for children 0 years - 6 months old. Since this value is obtained only for a small portion of the participant population, the decision was made to drop that column. 

There are some columns with comments to explain some of the missing data and the decision was made to keep those. The rest of the columns will be renamed to be more intuitive.

In [25]:
df_bmx.drop(['BMDSTATS','BMXHEAD','BMIHEAD'], axis = 1, inplace = True)

In [26]:
df_bmx.columns

Index(['SEQN', 'BMXWT', 'BMIWT', 'BMXRECUM', 'BMIRECUM', 'BMXHT', 'BMIHT',
       'BMXBMI', 'BMDBMIC', 'BMXLEG', 'BMILEG', 'BMXARML', 'BMIARML',
       'BMXARMC', 'BMIARMC', 'BMXWAIST', 'BMIWAIST', 'BMXHIP', 'BMIHIP'],
      dtype='object')

In [27]:
df_bmx = df_bmx.rename(columns={
    'SEQN':'participant_id',
    'BMXWT':'weight_kg',
    'BMIWT': 'weight_comment',
    'BMXRECUM': 'recumbent_length_cm',
    'BMIRECUM': 'recumbent_length_comment',
    'BMXHT': 'standing_height', 
    'BMIHT': 'standing_height_comment',
    'BMXBMI': 'bmi',
    'BMDBMIC': 'bmi_category_child',
    'BMXLEG': 'upper_leg_cm',
    'BMILEG': 'upper_leg_comment',
    'BMXARML': 'upper_arm_cm',
    'BMIARML': 'upper_arm_comment',
    'BMXARMC': 'arm_circumference_cm',
    'BMIARMC': 'arm_circ_comment',
    'BMXWAIST': 'waist_circ_cm',
    'BMIWAIST': 'waist_circ_comment', 
    'BMXHIP': 'hip_circ_cm', 
    'BMIHIP': 'hip_circ_comment'
})

In [28]:
df_bmx.head(10)

participant_id  weight_kg  weight_comment  recumbent_length_cm  \
1         109264.0       42.2             NaN                  NaN   
2         109265.0       12.0             NaN                 91.6   
3         109266.0       97.1             NaN                  NaN   
4         109269.0       13.6             NaN                 90.9   
5         109270.0       75.3             NaN                  NaN   
6         109271.0       98.8             NaN                  NaN   
7         109272.0        7.1             NaN                 63.6   
8         109273.0       74.3             NaN                  NaN   
9         109274.0      103.7             NaN                  NaN   
10        109275.0       20.9             NaN                  NaN   

    recumbent_length_comment  standing_height  standing_height_comment   bmi  \
1                        NaN            154.7                      NaN  17.6   
2                        NaN             89.3                      NaN  15.0   
3                        NaN            160.2                      NaN  37.8   
4                        NaN              NaN                      1.0   NaN   
5                        NaN            156.0                      NaN  30.9   
6                        NaN            182.3                      NaN  29.7   
7                        NaN              NaN                      NaN   NaN   
8                        NaN            184.2                      NaN  21.9   
9                        NaN            185.3                      NaN  30.2   
10                       NaN            120.4                      NaN  14.4   

    bmi_category_child  upper_leg_cm  upper_leg_comment  upper_arm_cm  \
1                  2.0          36.3                NaN          33.8   
2                  2.0           NaN                NaN          18.6   
3                  NaN          40.8                NaN          34.7   
4                  NaN           NaN                NaN           NaN   
5                  4.0          42.6                NaN          36.1   
6                  NaN          40.1                NaN          42.0   
7                  NaN           NaN                NaN          13.0   
8                  NaN          41.0                NaN          41.1   
9                  NaN          44.0                NaN          47.0   
10                 2.0           NaN                NaN           NaN   

    upper_arm_comment  arm_circumference_cm  arm_circ_comment  waist_circ_cm  \
1                 NaN                  22.7               NaN           63.8   
2                 NaN                  14.8               NaN           41.2   
3                 NaN                  35.8               NaN          117.9   
4                 1.0                   NaN               1.0            NaN   
5                 NaN                  31.0               NaN           91.4   
6                 NaN                  38.2               NaN          120.4   
7                 NaN                  15.5               NaN            NaN   
8                 NaN                  30.2               NaN           86.8   
9                 NaN                  32.0               NaN          109.6   
10                1.0                   NaN               1.0            NaN   

    waist_circ_comment  hip_circ_cm  hip_circ_comment  
1                  NaN         85.0               NaN  
2                  NaN          NaN               NaN  
3                  NaN        126.1               NaN  
4                  1.0          NaN               NaN  
5                  NaN          NaN               NaN  
6                  NaN        108.2               NaN  
7                  NaN          NaN               NaN  
8                  NaN         94.5               NaN  
9                  NaN        107.8               NaN  
10                 1.0          NaN               NaN

With the comment sections, most of the sections are marked "Could not obtain". The comment sections that only have such notation was removed, as the lack of value is self evident with missing values (NaN).

In [29]:
df_bmx.drop(['recumbent_length_comment','upper_leg_comment','upper_arm_comment',
         'arm_circ_comment','waist_circ_comment','hip_circ_comment'], axis = 1, inplace = True)

In [30]:
df_bmx.head(10)

participant_id  weight_kg  weight_comment  recumbent_length_cm  \
1         109264.0       42.2             NaN                  NaN   
2         109265.0       12.0             NaN                 91.6   
3         109266.0       97.1             NaN                  NaN   
4         109269.0       13.6             NaN                 90.9   
5         109270.0       75.3             NaN                  NaN   
6         109271.0       98.8             NaN                  NaN   
7         109272.0        7.1             NaN                 63.6   
8         109273.0       74.3             NaN                  NaN   
9         109274.0      103.7             NaN                  NaN   
10        109275.0       20.9             NaN                  NaN   

    standing_height  standing_height_comment   bmi  bmi_category_child  \
1             154.7                      NaN  17.6                 2.0   
2              89.3                      NaN  15.0                 2.0   
3             160.2                      NaN  37.8                 NaN   
4               NaN                      1.0   NaN                 NaN   
5             156.0                      NaN  30.9                 4.0   
6             182.3                      NaN  29.7                 NaN   
7               NaN                      NaN   NaN                 NaN   
8             184.2                      NaN  21.9                 NaN   
9             185.3                      NaN  30.2                 NaN   
10            120.4                      NaN  14.4                 2.0   

    upper_leg_cm  upper_arm_cm  arm_circumference_cm  waist_circ_cm  \
1           36.3          33.8                  22.7           63.8   
2            NaN          18.6                  14.8           41.2   
3           40.8          34.7                  35.8          117.9   
4            NaN           NaN                   NaN            NaN   
5           42.6          36.1                  31.0           91.4   
6           40.1          42.0                  38.2          120.4   
7            NaN          13.0                  15.5            NaN   
8           41.0          41.1                  30.2           86.8   
9           44.0          47.0                  32.0          109.6   
10           NaN           NaN                   NaN            NaN   

    hip_circ_cm  
1          85.0  
2           NaN  
3         126.1  
4           NaN  
5           NaN  
6         108.2  
7           NaN  
8          94.5  
9         107.8  
10          NaN

In [31]:
df_bmx.shape

(14107, 13)

The comment sections that are kept have values more than just "Not Obtained". These will be interpreted with appropriate comment and translated into string values

In [32]:
df_bmx.to_csv('cleaned_bodymeasures.csv', index=False)

### Lab Results Data Cleaning

Laboratory analysis is a vital component of the NHANES 2017–2020 dataset, offering detailed biomarker data that can be leveraged for predictive modeling. The most common sources of lab data are blood and urine samples. To facilitate data cleaning and relational database design, all lab results were categorized into two main groups: blood-based tests and urine-based tests.

This separation enhances the clarity and usability of the database structure, especially in the context of developing an obesity prediction model that emphasizes blood biomarkers. While blood tests serve as the primary focus due to their strong association with metabolic health, urine tests are also retained for potential secondary insights.

During data cleaning, each lab dataset from the 2017–2020 cycle was examined for completeness, consistent formatting, and variable alignment. Efforts were made to standardize variable names and units across different files to ensure compatibility. Only participants with valid lab results and complete demographic data were retained for analysis.

#### Urine Labs

##### Albumin and Creatinine

Albumin is a protein often found in urine if there is damage to the kidneys. Creatinine is waste product that is excreted in urine and helps evaluate the kidney function.

In [33]:
file_path = '2017-2020/urine/1.P_ALB_CR.xpt'

df, meta = pyreadstat.read_xport(file_path)

In [34]:
df.shape

(13027, 8)

In [35]:
df.head(10)

SEQN  URXUMA  URXUMS  URDUMALC  URXUCR   URXCRS  URDUCRLC  URDACT
0  109264.0     NaN     NaN       NaN     NaN      NaN       NaN     NaN
1  109266.0     5.5     5.5       0.0    36.0   3182.4       0.0   15.28
2  109270.0     4.0     4.0       0.0   165.0  14586.0       0.0    2.42
3  109271.0     2.4     2.4       0.0    32.0   2828.8       0.0    7.50
4  109273.0     4.9     4.9       0.0   121.0  10696.4       0.0    4.05
5  109274.0    12.8    12.8       0.0   120.0  10608.0       0.0   10.67
6  109275.0     3.7     3.7       0.0    20.0   1768.0       0.0   18.50
7  109277.0    14.7    14.7       0.0   244.0  21569.6       0.0    6.02
8  109278.0     8.4     8.4       0.0   124.0  10961.6       0.0    6.77
9  109279.0    13.9    13.9       0.0   251.0  22188.4       0.0    5.54

In [36]:
df.columns

Index(['SEQN', 'URXUMA', 'URXUMS', 'URDUMALC', 'URXUCR', 'URXCRS', 'URDUCRLC',
       'URDACT'],
      dtype='object')

In [37]:
df = df.rename(columns={
    'SEQN':'participant_id',
    'URXUMA': 'albumin_urine_ug_mL',
    'URXUMS': 'albumin_urine_mg_L', 
    'URDUMALC': 'alb_comment',
    'URXUCR': 'creatinine_urine_mg_dL', 
    'URXCRS': 'creatinine_urine_umol_L', 
    'URDUCRLC': 'creatinine_comment',
    'URDACT': 'alb_creat_ratio'
})

In [38]:
df.isnull().sum()

participant_id               0
albumin_urine_ug_mL        517
albumin_urine_mg_L         517
alb_comment                517
creatinine_urine_mg_dL     518
creatinine_urine_umol_L    518
creatinine_comment         518
alb_creat_ratio            518
dtype: int64

Some columns contain a similar number of missing (NaN) values, raising the question of whether these missing values occur for the same participants. To assess potential overlap, the following code was executed to identify if the missing values correspond to the same participant IDs. This check helps ensure that removing rows with missing data will not disproportionately reduce the dataset.

In [39]:
# Assuming your dataframe is called df, and participant ID column is 'SEQN'
def get_common_nan_ids(df, col1, col2, id_col='participant_id', verbose=True):
    """
    Returns a set of participant IDs where BOTH col1 and col2 are NaN.
    
    Parameters:
    - df: pandas DataFrame
    - col1, col2: column names to check for NaNs
    - id_col: column name for participant IDs (default 'participant_id')
    
    Returns:
    - Set of participant IDs with NaNs in both columns
    """
    ids_nan_col1 = set(df.loc[df[col1].isna(), id_col])
    ids_nan_col2 = set(df.loc[df[col2].isna(), id_col])
    common_nan_ids = ids_nan_col1.intersection(ids_nan_col2)
    
    if verbose:
        print(f"Number of NaNs in {col1}: {len(ids_nan_col1)}")
        print(f"Number of NaNs in {col2}: {len(ids_nan_col2)}")
        print(f"Number of IDs with NaNs in both columns: {len(common_nan_ids)}")
    
    return common_nan_ids

In [40]:
def drop_rows_with_common_nan_ids(df, col1, col2, id_col='participant_id'):
    """
    Drops rows where BOTH col1 and col2 are NaN.
    Uses get_common_nan_ids() to identify rows.
    
    Parameters:
    - df: pandas DataFrame
    - col1, col2: column names to check for NaNs
    - id_col: column name for participant IDs (default 'participant_id')
    
    Returns:
    - cleaned DataFrame (copy)
    """
    common_nan_ids = get_common_nan_ids(df, col1, col2, id_col, verbose=False)
    rows_dropped = df[id_col].isin(common_nan_ids).sum()
    
    print(f"Rows dropped where both {col1} and {col2} were NaN: {rows_dropped}")
    
    return df[~(df[col1].isna() & df[col2].isna())].copy()

In [41]:
common_nan = get_common_nan_ids(df,'albumin_urine_ug_mL','creatinine_urine_mg_dL')

Number of NaNs in albumin_urine_ug_mL: 517
Number of NaNs in creatinine_urine_mg_dL: 518
Number of IDs with NaNs in both columns: 517


In [42]:
df_cleaned = drop_rows_with_common_nan_ids(df, 'albumin_urine_ug_mL', 'creatinine_urine_mg_dL')

Rows dropped where both albumin_urine_ug_mL and creatinine_urine_mg_dL were NaN: 517


##### Arsenic

Arsenic is a naturally occurring mineral present in water, air, and soil, existing in both organic and inorganic forms. While the inorganic form is more toxic, exposure to any form of arsenic can be harmful to human health. Elevated arsenic levels have been linked to adverse effects on multiple body systems, including the cardiovascular and endocrine systems [3].

Urinary arsenic levels serve as a biomarker reflecting the concentration of arsenic in the bloodstream. Given arsenic’s potential impact on the endocrine system—a key regulator of metabolism and body weight—these lab values are relevant for investigating associations with obesity in this analysis.

##### Total Arsenic

In [43]:
file_path = '2017-2020/urine/2.P_UTAS.xpt'

df1, meta = pyreadstat.read_xport(file_path)

In [44]:
df1.head(10)

SEQN       WTSAPRP  URXUAS  URDUASLC
0  109266.0  28660.015986    2.05       0.0
1  109270.0  17900.682903    3.25       0.0
2  109273.0  80106.859617    5.16       0.0
3  109274.0  24512.276280    2.92       0.0
4  109287.0  25828.523003    5.80       0.0
5  109288.0   8535.018174    4.24       0.0
6  109290.0  12410.268374  144.72       0.0
7  109295.0  28235.246814    3.48       0.0
8  109300.0  66737.887353    5.08       0.0
9  109309.0  33019.729726    0.95       0.0

In [45]:
df1.columns

Index(['SEQN', 'WTSAPRP', 'URXUAS', 'URDUASLC'], dtype='object')

In [46]:
df1 = df1.rename(columns={
    'SEQN':'participant_id',
    'URXUAS':'total_arsenic_ug_L',
    'URDUASLC': 'total_arsenic_comment'
})

In [47]:
df1.drop('WTSAPRP', axis=1,inplace = True)

In [48]:
df1.isna().sum()

participant_id             0
total_arsenic_ug_L       320
total_arsenic_comment    320
dtype: int64

In [49]:
df1.dropna(axis=1, inplace=True)

##### Speciated Arsenic

In [50]:
file_path = '2017-2020/urine/3.P_UAS.xpt'

df2, meta = pyreadstat.read_xport(file_path)

In [51]:
df2.shape

(4890, 14)

In [52]:
df2.columns

Index(['SEQN', 'WTSAPRP', 'URXUAS3', 'URDUA3LC', 'URXUAS5', 'URDUA5LC',
       'URXUAB', 'URDUABLC', 'URXUAC', 'URDUACLC', 'URXUDMA', 'URDUDALC',
       'URXUMMA', 'URDUMMAL'],
      dtype='object')

In [53]:
df2 = df2.drop('WTSAPRP', axis = 1)

In [54]:
df2 = df2.rename(columns={
    'SEQN':'participant_id',
    'WTSAPRP': 'subsample_weight',
    'URXUAS3':'arsenous_acid_ug_L',
    'URDUA3LC':'arsenous_acid_comment', 
    'URXUAS5': 'arsenic_acid_ug_L',
    'URDUA5LC': 'arsenic_acid_comment',
    'URXUAB': 'arsenobetaine_ug_L',
    'URDUABLC':'arsenobetaine_comment',
    'URXUAC': 'arsenocholoine_ug_L',
    'URDUACLC': 'arsenocholine_comment',
    'URXUDMA': 'dimethylarsinic_acid_ug_L', 
    'URDUDALC': 'dimethylarsinic_comment',
    'URXUMMA': 'monomethylarsonic_acid_ug_L',
    'URDUMMAL': 'monometylarsonic_comment'
})

In [55]:
df2.head(10)

participant_id  arsenous_acid_ug_L  arsenous_acid_comment  \
0        109266.0                0.08                    1.0   
1        109270.0                0.51                    0.0   
2        109273.0                0.08                    1.0   
3        109274.0                0.08                    1.0   
4        109287.0                0.08                    1.0   
5        109288.0                0.08                    1.0   
6        109290.0                0.44                    0.0   
7        109295.0                0.08                    1.0   
8        109300.0                0.08                    1.0   
9        109309.0                0.08                    1.0   

   arsenic_acid_ug_L  arsenic_acid_comment  arsenobetaine_ug_L  \
0               0.56                   1.0                0.82   
1               0.56                   1.0                0.82   
2               0.56                   1.0                2.74   
3               0.56                   1.0                0.82   
4               0.56                   1.0                0.82   
5               0.56                   1.0                0.82   
6               0.56                   1.0              147.82   
7               0.56                   1.0                1.28   
8               0.56                   1.0                3.03   
9               0.56                   1.0                0.82   

   arsenobetaine_comment  arsenocholoine_ug_L  arsenocholine_comment  \
0                    1.0                 0.08                    1.0   
1                    1.0                 0.08                    1.0   
2                    0.0                 0.08                    1.0   
3                    1.0                 0.08                    1.0   
4                    1.0                 0.08                    1.0   
5                    1.0                 0.08                    1.0   
6                    0.0                 0.76                    0.0   
7                    0.0                 0.08                    1.0   
8                    0.0                 0.08                    1.0   
9                    1.0                 0.08                    1.0   

   dimethylarsinic_acid_ug_L  dimethylarsinic_comment  \
0                       1.35                      1.0   
1                       2.07                      0.0   
2                       1.35                      1.0   
3                       1.35                      1.0   
4                       3.60                      0.0   
5                       4.22                      0.0   
6                       6.63                      0.0   
7                       1.35                      1.0   
8                       1.35                      1.0   
9                       1.35                      1.0   

   monomethylarsonic_acid_ug_L  monometylarsonic_comment  
0                         0.14                       1.0  
1                         0.14                       1.0  
2                         0.14                       1.0  
3                         0.14                       1.0  
4                         0.14                       1.0  
5                         0.14                       1.0  
6                         0.70                       0.0  
7                         0.14                       1.0  
8                         0.14                       1.0  
9                         0.14                       1.0

In [56]:
df2.isna().sum()

participant_id                   0
arsenous_acid_ug_L             265
arsenous_acid_comment          265
arsenic_acid_ug_L              265
arsenic_acid_comment           265
arsenobetaine_ug_L             265
arsenobetaine_comment          265
arsenocholoine_ug_L            265
arsenocholine_comment          265
dimethylarsinic_acid_ug_L      265
dimethylarsinic_comment        265
monomethylarsonic_acid_ug_L    265
monometylarsonic_comment       265
dtype: int64

In [57]:
df2_cleaned = drop_rows_with_common_nan_ids(df2, 'arsenous_acid_ug_L', 'arsenic_acid_ug_L')

Rows dropped where both arsenous_acid_ug_L and arsenic_acid_ug_L were NaN: 265


##### Chromium

Chromium is a trace mineral found in two main forms: trivalent chromium (Cr³⁺), which is naturally present in food and supplements, and hexavalent chromium (Cr⁶⁺), a toxic by-product of industrial processes such as metal manufacturing 
[4]. Trivalent chromium has been suggested to play a role in carbohydrate, lipid, and protein metabolism by enhancing insulin action; however, no definitive physiological function has been firmly established. In contrast, hexavalent chromium is recognized as potentially carcinogenic, especially when inhaled or ingested in high amounts. Due to the uncertainty surrounding its physiological role, a standardized reference range for chromium has not been established.

For the purposes of this project, which focuses on assessing correlations with obesity, trivalent chromium is of particular interest due to its possible metabolic effects. However, a limitation of the NHANES dataset is that urinary chromium measurements are not specifically categorized by valence state. While this makes it difficult to distinguish between beneficial and harmful forms, urine chromium concentration can still serve as a proxy for overall chromium exposure. Monitoring these levels may offer insight into whether exposure levels are within a potentially physiological or harmful range.

In [58]:
file_path = '2017-2020/urine/4.P_UCM.xpt'

df3, meta = pyreadstat.read_xport(file_path)

In [59]:
df3.columns

Index(['SEQN', 'WTSAPRP', 'URXUCM', 'URDUCMLC'], dtype='object')

In [60]:
df3 = df3.drop('WTSAPRP',axis=1)

In [61]:
df3.head(10)

SEQN  URXUCM  URDUCMLC
0  109266.0    0.13       1.0
1  109270.0    0.13       1.0
2  109273.0    0.19       0.0
3  109274.0    0.40       0.0
4  109287.0    0.13       1.0
5  109288.0    0.29       0.0
6  109290.0    0.27       0.0
7  109295.0    0.26       0.0
8  109300.0    0.13       1.0
9  109309.0    0.13       1.0

In [62]:
df3.isna().sum()

SEQN          0
URXUCM      321
URDUCMLC    321
dtype: int64

Since there is only one lab value that is being measured in this dataset, any NaN values will be dropped.

In [63]:
df3 = df3.rename(columns={
    'SEQN':'participant_id',
   'URXUCM': 'chromium_ug_L',
    'URDUCMLC': 'chromium_comment'
})

In [64]:
df3.dropna(axis=1,inplace=True)

##### Flame Retardant

Flame retardants (FRs) are chemicals applied to materials such as furniture, electronics, electrical devices, and construction products to reduce flammability and slow the spread of fire. There are several major classes of flame retardants, including brominated flame retardants (BFRs), hexabromocyclododecane (HBCD), organophosphate flame retardants (OPFRs), tetrabromobisphenol A (TBBPA), and polybrominated diphenyl ethers (PBDEs). These compounds are highly persistent in the environment and can accumulate in human tissue due to their resistance to degradation [5].

Among them, PBDEs have been extensively studied for their potential adverse health effects, including endocrine and thyroid disruption, immunotoxicity, reproductive toxicity, carcinogenicity, and negative impacts on fetal and child development. Although young children are particularly vulnerable to these effects, adults are also susceptible to long-term exposure. Given their endocrine-disrupting properties, flame retardants may play a role in the development or progression of obesity.

According to the documentation from NHANES for this lab, the unit of measurement is all ng/mL.

In [65]:
file_path = '2017-2020/urine/5.P_FR.xpt'

df4, meta = pyreadstat.read_xport(file_path)

In [66]:
df4.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4929 entries, 0 to 4928
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   SEQN      4929 non-null   float64
 1   WTSBPRP   4929 non-null   float64
 2   URXBCPP   4617 non-null   float64
 3   URDBCPLC  4617 non-null   float64
 4   URXBCEP   4618 non-null   float64
 5   URDCEPLC  4618 non-null   float64
 6   URXBDCP   4599 non-null   float64
 7   URDBDCLC  4599 non-null   float64
 8   URXDBUP   4614 non-null   float64
 9   URDDUPLC  4614 non-null   float64
 10  URXDPHP   4622 non-null   float64
 11  URDDPHLC  4622 non-null   float64
 12  URXTBBA   4622 non-null   float64
 13  URDBBALC  4622 non-null   float64
dtypes: float64(14)
memory usage: 539.2 KB


In [67]:
df4.columns

Index(['SEQN', 'WTSBPRP', 'URXBCPP', 'URDBCPLC', 'URXBCEP', 'URDCEPLC',
       'URXBDCP', 'URDBDCLC', 'URXDBUP', 'URDDUPLC', 'URXDPHP', 'URDDPHLC',
       'URXTBBA', 'URDBBALC'],
      dtype='object')

In [68]:
df4.head(10)

SEQN       WTSBPRP  URXBCPP  URDBCPLC  URXBCEP  URDCEPLC  URXBDCP  \
0  109271.0  20156.439742   0.0707       1.0   0.0707       1.0   0.1370   
1  109277.0  51738.369518   0.7090       0.0   0.0707       1.0   1.1500   
2  109282.0  97190.554500      NaN       NaN   4.2300       0.0   3.6600   
3  109285.0  85548.221421   0.0707       1.0   4.4200       0.0   2.2700   
4  109288.0   9103.868995   0.3470       0.0   0.1450       0.0   0.8590   
5  109301.0  14215.937370   0.1160       0.0   0.0707       1.0   5.1800   
6  109302.0   5984.497323   0.0707       1.0   0.0707       1.0   4.8100   
7  109303.0  16549.099643   0.0707       1.0   0.3660       0.0   0.0707   
8  109304.0  40089.354988   0.1740       0.0   0.3860       0.0   2.6800   
9  109307.0  49745.101247   0.0707       1.0   0.0707       1.0   0.9030   

   URDBDCLC  URXDBUP  URDDUPLC  URXDPHP  URDDPHLC  URXTBBA  URDBBALC  
0       0.0   0.0707       1.0    0.886       0.0   0.0354       1.0  
1       0.0   0.1380       0.0    1.300       0.0   0.0354       1.0  
2       0.0   0.5010       0.0    2.050       0.0   0.0354       1.0  
3       0.0   0.0707       1.0   10.300       0.0   0.0620       0.0  
4       0.0   0.0707       1.0    1.290       0.0   0.0354       1.0  
5       0.0   0.4500       0.0    0.701       0.0   0.0354       1.0  
6       0.0   0.1710       0.0    0.732       0.0   0.0354       1.0  
7       1.0   0.0707       1.0    0.162       0.0   0.0354       1.0  
8       0.0   0.1400       0.0    1.280       0.0   0.0354       1.0  
9       0.0   0.0707       1.0    1.550       0.0   0.1210       0.0

In [69]:
df4 = df4.drop('WTSBPRP', axis = 1)

In [70]:
df4 = df4.rename(columns={
    'SEQN': 'participant_id',
    'URXBCPP': '1_chloro_2_propyl_phosphate', 
    'URDBCPLC' : '1ch_2pro_comment', 
    'URXBCEP': 'bis_1_chloroethyl_phosphate',
    'URDCEPLC' : 'bis_1_chlo_phos_comment',
    'URXBDCP': '1_3_dichloro_2_propyl_phosphate', 
    'URDBDCLC': '1_3_di_2_pro_comment', 
    'URXDBUP': 'dibutyl_phosphate',
    'URDDUPLC': 'dibutyl_phos_comment', 
    'URXDPHP': 'diphenyl_phosphate',
    'URDDPHLC': 'diphe_phos_comment',
    'URXTBBA': '2_3_4_5_tetrabromobenzoic_acid',
    'URDBBALC': '2_3_4_5_tet_comment'
})

In [71]:
df4.isnull().sum()

participant_id                       0
1_chloro_2_propyl_phosphate        312
1ch_2pro_comment                   312
bis_1_chloroethyl_phosphate        311
bis_1_chlo_phos_comment            311
1_3_dichloro_2_propyl_phosphate    330
1_3_di_2_pro_comment               330
dibutyl_phosphate                  315
dibutyl_phos_comment               315
diphenyl_phosphate                 307
diphe_phos_comment                 307
2_3_4_5_tetrabromobenzoic_acid     307
2_3_4_5_tet_comment                307
dtype: int64

There are multiple columns that have over 300 NaN values. There may be rows that are missing all of these values but there may be rows that are missing some of the values. The decision was made to drop rows that are missing all values to ensure retention of meaningful data.

In [72]:
df4.columns #to make it easy to copy and paste the names of these columns without risking any typos

Index(['participant_id', '1_chloro_2_propyl_phosphate', '1ch_2pro_comment',
       'bis_1_chloroethyl_phosphate', 'bis_1_chlo_phos_comment',
       '1_3_dichloro_2_propyl_phosphate', '1_3_di_2_pro_comment',
       'dibutyl_phosphate', 'dibutyl_phos_comment', 'diphenyl_phosphate',
       'diphe_phos_comment', '2_3_4_5_tetrabromobenzoic_acid',
       '2_3_4_5_tet_comment'],
      dtype='object')

In [73]:
fr_value_cols = [
    '1_chloro_2_propyl_phosphate',
    'bis_1_chloroethyl_phosphate',
    '1_3_dichloro_2_propyl_phosphate',
    'dibutyl_phosphate',
    'diphenyl_phosphate',
    '2_3_4_5_tetrabromobenzoic_acid'
]

rows_all_nan = df4[fr_value_cols].isna().all(axis=1)
rows_missing_all_fr = df4[rows_all_nan]

print(f"Number of rows missing all FR values: {rows_all_nan.sum()}")

Number of rows missing all FR values: 307


In [74]:
df4_clean = df4.dropna(how='all', subset=fr_value_cols)

In [75]:
df4_clean.head(10)

participant_id  1_chloro_2_propyl_phosphate  1ch_2pro_comment  \
0        109271.0                       0.0707               1.0   
1        109277.0                       0.7090               0.0   
2        109282.0                          NaN               NaN   
3        109285.0                       0.0707               1.0   
4        109288.0                       0.3470               0.0   
5        109301.0                       0.1160               0.0   
6        109302.0                       0.0707               1.0   
7        109303.0                       0.0707               1.0   
8        109304.0                       0.1740               0.0   
9        109307.0                       0.0707               1.0   

   bis_1_chloroethyl_phosphate  bis_1_chlo_phos_comment  \
0                       0.0707                      1.0   
1                       0.0707                      1.0   
2                       4.2300                      0.0   
3                       4.4200                      0.0   
4                       0.1450                      0.0   
5                       0.0707                      1.0   
6                       0.0707                      1.0   
7                       0.3660                      0.0   
8                       0.3860                      0.0   
9                       0.0707                      1.0   

   1_3_dichloro_2_propyl_phosphate  1_3_di_2_pro_comment  dibutyl_phosphate  \
0                           0.1370                   0.0             0.0707   
1                           1.1500                   0.0             0.1380   
2                           3.6600                   0.0             0.5010   
3                           2.2700                   0.0             0.0707   
4                           0.8590                   0.0             0.0707   
5                           5.1800                   0.0             0.4500   
6                           4.8100                   0.0             0.1710   
7                           0.0707                   1.0             0.0707   
8                           2.6800                   0.0             0.1400   
9                           0.9030                   0.0             0.0707   

   dibutyl_phos_comment  diphenyl_phosphate  diphe_phos_comment  \
0                   1.0               0.886                 0.0   
1                   0.0               1.300                 0.0   
2                   0.0               2.050                 0.0   
3                   1.0              10.300                 0.0   
4                   1.0               1.290                 0.0   
5                   0.0               0.701                 0.0   
6                   0.0               0.732                 0.0   
7                   1.0               0.162                 0.0   
8                   0.0               1.280                 0.0   
9                   1.0               1.550                 0.0   

   2_3_4_5_tetrabromobenzoic_acid  2_3_4_5_tet_comment  
0                          0.0354                  1.0  
1                          0.0354                  1.0  
2                          0.0354                  1.0  
3                          0.0620                  0.0  
4                          0.0354                  1.0  
5                          0.0354                  1.0  
6                          0.0354                  1.0  
7                          0.0354                  1.0  
8                          0.0354                  1.0  
9                          0.1210                  0.0

In [76]:
file_path = '2017-2020/urine/6.P_SSFR.xpt'

df5, meta = pyreadstat.read_xport(file_path)

In [77]:
df5.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4929 entries, 0 to 4928
Data columns (total 6 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   SEQN     4929 non-null   float64
 1   WTSSBPP  4929 non-null   float64
 2   SSIPPP   3913 non-null   float64
 3   SSIPPPL  3913 non-null   float64
 4   SSBPPP   3923 non-null   float64
 5   SSBPPPL  3923 non-null   float64
dtypes: float64(6)
memory usage: 231.2 KB


In [78]:
df5.isnull().sum()

SEQN          0
WTSSBPP       0
SSIPPP     1016
SSIPPPL    1016
SSBPPP     1006
SSBPPPL    1006
dtype: int64

In [79]:
df5 = df5.drop('WTSSBPP', axis=1)

In [80]:
df5 = df5.rename(columns={
    'SEQN': 'participant_id',
    'SSIPPP' : '2_isopropylphenyl_phenyl_phosphate',
    'SSIPPPL' : '2_isopropylphenyl_phenyl_phosphate_comment',
    'SSBPPP' : '4_tert_butylphenyl_phenyl_phosphate',
    'SSBPPPL' : '4_tert_butylphenyl_phenyl_phosphate_comment'
})


In [81]:
common_nan = get_common_nan_ids(df5, '2_isopropylphenyl_phenyl_phosphate', '4_tert_butylphenyl_phenyl_phosphate', id_col='participant_id')

Number of NaNs in 2_isopropylphenyl_phenyl_phosphate: 1016
Number of NaNs in 4_tert_butylphenyl_phenyl_phosphate: 1006
Number of IDs with NaNs in both columns: 1006


In [82]:
df5_cleaned = drop_rows_with_common_nan_ids(df5, '2_isopropylphenyl_phenyl_phosphate', '4_tert_butylphenyl_phenyl_phosphate', id_col='participant_id')

Rows dropped where both 2_isopropylphenyl_phenyl_phosphate and 4_tert_butylphenyl_phenyl_phosphate were NaN: 1006


##### Iodine

Iodine is a trace element commonly found in foods and iodized salt, playing a critical role in thyroid function. The thyroid gland is essential for regulating metabolism and is particularly important for fetal and infant development [6].

Approximately 90% of ingested iodine is excreted in the urine. While urinary iodine concentration is not considered a reliable indicator of iodine status at the individual level, it can be used for assessing iodine sufficiency across populations [7].

In [83]:
file_path = '2017-2020/urine/7.P_UIO.xpt'

df6, meta = pyreadstat.read_xport(file_path)

In [84]:
df6.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4890 entries, 0 to 4889
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   SEQN      4890 non-null   float64
 1   WTSAPRP   4890 non-null   float64
 2   URXUIO    4600 non-null   float64
 3   URDUIOLC  4600 non-null   float64
dtypes: float64(4)
memory usage: 152.9 KB


In [85]:
df6 = df6.drop('WTSAPRP', axis = 1)

In [86]:
df6 = df6.rename(columns={
    'SEQN' : 'participant_id',
    'URXUIO' : 'urine_iodine',
    'URDUIOLC' : 'urine_iodine_comment'
})

In [87]:
df6.isna().sum()

participant_id            0
urine_iodine            290
urine_iodine_comment    290
dtype: int64

In [88]:
df6_clean = df6.dropna(subset = ['urine_iodine'])

##### Mercury

Mercury is a heavy metal historically used in devices such as barometers and thermometers. At elevated levels, it is known to cause neurotoxicity, with particularly severe effects on fetal development. Environmental exposure—especially in occupational settings involving manufacturing or chemical production—is a common source of mercury-related toxicity.

Urinary mercury concentration is a standard method for assessing inorganic mercury exposure. Clinical symptoms may begin to appear at concentrations around 100 µg/L, and levels exceeding 800 µg/L can be fatal [9].

In [89]:
file_path = '2017-2020/urine/8.P_UHG.xpt'

df7, meta = pyreadstat.read_xport(file_path)

In [90]:
df7.head(10)

SEQN       WTSAPRP  URXUHG  URDUHGLC
0  109266.0  28660.015986    0.09       1.0
1  109270.0  17900.682903    0.09       1.0
2  109273.0  80106.859617    0.09       1.0
3  109274.0  24512.276280    0.09       1.0
4  109287.0  25828.523003    0.09       1.0
5  109288.0   8535.018174    0.09       1.0
6  109290.0  12410.268374    1.27       0.0
7  109295.0  28235.246814    0.09       1.0
8  109300.0  66737.887353    0.09       1.0
9  109309.0  33019.729726    0.09       1.0

In [91]:
df7.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4890 entries, 0 to 4889
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   SEQN      4890 non-null   float64
 1   WTSAPRP   4890 non-null   float64
 2   URXUHG    4600 non-null   float64
 3   URDUHGLC  4600 non-null   float64
dtypes: float64(4)
memory usage: 152.9 KB


In [92]:
df7 = df7.drop('WTSAPRP', axis =1)

In [93]:
df7 = df7.rename(columns={
    'SEQN': 'participant_id',
    'URXUHG' : 'urine_mercury',
    'URDUHGLC' : 'urine_mercury_comment'
})

In [94]:
df7_clean = df7.dropna(subset=['urine_mercury'])

In [95]:
df7_clean.head(5)

participant_id  urine_mercury  urine_mercury_comment
0        109266.0           0.09                    1.0
1        109270.0           0.09                    1.0
2        109273.0           0.09                    1.0
3        109274.0           0.09                    1.0
4        109287.0           0.09                    1.0

##### Metals

There are other types of metals that can be found in urine. From this dataset from NHANES, the metals tested are: barium, cadmium, cobalt, cesium, molybdenum, manganese, lead, antiomny, tin, thallium, and tungsten. 

Barium, if consumed in high concentrations, it can cause cardiac arrhythmias or paralysis [8]

Cadmium is a known carcinogen which is excreted primarily by the renal system. Cadminum can lead to various health conditions throughout the body including kidney and liver dysfunctions. [10]




In [96]:
file_path = '2017-2020/urine/9.P_UM.xpt'

df8, meta = pyreadstat.read_xport(file_path)

In [97]:
df8.columns.to_list()

['SEQN',
 'WTSAPRP',
 'URXUBA',
 'URDUBALC',
 'URXUCD',
 'URDUCDLC',
 'URXUCO',
 'URDUCOLC',
 'URXUCS',
 'URDUCSLC',
 'URXUMO',
 'URDUMOLC',
 'URXUMN',
 'URDUMNLC',
 'URXUPB',
 'URDUPBLC',
 'URXUSB',
 'URDUSBLC',
 'URXUSN',
 'URDUSNLC',
 'URXUTL',
 'URDUTLLC',
 'URXUTU',
 'URDUTULC']

In [98]:
df8 = df8.rename(columns = {
    'SEQN': 'participant_id',
    'URXUBA': 'urine_barium',
    'URDUBALC': 'barium_comment',
    'URXUCD': 'urine_cadmium',
    'URDUCDLC': 'cadmium_comment',
    'URXUCO': 'urine_cobalt',
    'URDUCOLC': 'cobalt_comment',
     'URXUCS': 'urine_cesium',
     'URDUCSLC': 'cesium_comment',
     'URXUMO': 'urine_molybdenum',
     'URDUMOLC': 'molybdenum_comment',
     'URXUMN': 'urine_manganese',
     'URDUMNLC':'manganese_comment',
     'URXUPB':'urine_lead',
     'URDUPBLC':'lead_comment',
     'URXUSB':'urine_antimony',
     'URDUSBLC':'antimony_comment',
     'URXUSN':'urine_tin',
     'URDUSNLC':'tin_comment',
     'URXUTL':'urine_thallium',
     'URDUTLLC':'thallium_comment',
     'URXUTU':'urine_tungsten',
     'URDUTULC':'tungsten_comment'
})

In [99]:
df8 = df8.drop('WTSAPRP', axis = 1)

In [100]:
df8.head(10)

participant_id  urine_barium  barium_comment  urine_cadmium  \
0        109266.0         0.359             0.0          0.039   
1        109270.0         2.422             0.0          0.868   
2        109273.0         0.370             0.0          0.213   
3        109274.0         1.720             0.0          0.184   
4        109287.0         7.531             0.0          0.215   
5        109288.0         0.271             0.0          0.039   
6        109290.0         0.720             0.0          0.631   
7        109295.0         1.270             0.0          0.039   
8        109300.0         0.690             0.0          0.166   
9        109309.0         0.610             0.0          0.039   

   cadmium_comment  urine_cobalt  cobalt_comment  urine_cesium  \
0              1.0         0.214             0.0         2.160   
1              0.0         0.449             0.0         9.640   
2              0.0         0.274             0.0         2.330   
3              0.0         0.482             0.0         2.803   
4              0.0         0.303             0.0         2.850   
5              1.0         0.097             0.0         5.980   
6              0.0         0.379             0.0        11.910   
7              1.0         0.323             0.0         4.145   
8              0.0         0.171             0.0         2.129   
9              1.0         0.263             0.0         1.673   

   cesium_comment  urine_molybdenum  ...  urine_lead  lead_comment  \
0             0.0              8.00  ...       0.170           0.0   
1             0.0             78.66  ...       0.532           0.0   
2             0.0             33.09  ...       0.280           0.0   
3             0.0             74.82  ...       0.300           0.0   
4             0.0            142.88  ...       0.244           0.0   
5             0.0             36.21  ...         NaN           NaN   
6             0.0            111.07  ...       0.641           0.0   
7             0.0             25.76  ...       0.060           0.0   
8             0.0             15.42  ...       0.130           0.0   
9             0.0              5.14  ...         NaN           NaN   

   urine_antimony  antimony_comment  urine_tin  tin_comment  urine_thallium  \
0           0.016               1.0       0.14          1.0           0.064   
1           0.046               0.0       2.35          0.0           0.354   
2           0.082               0.0       0.14          1.0           0.078   
3           0.053               0.0       4.09          0.0           0.159   
4           0.092               0.0       3.91          0.0           0.161   
5           0.067               0.0       0.97          0.0           0.175   
6           0.116               0.0       1.31          0.0           0.577   
7           0.022               0.0       0.14          1.0           0.147   
8           0.016               1.0       0.14          1.0           0.063   
9           0.028               0.0       0.30          0.0           0.099   

   thallium_comment  urine_tungsten  tungsten_comment  
0               0.0           0.013               1.0  
1               0.0           0.271               0.0  
2               0.0           0.036               0.0  
3               0.0           0.099               0.0  
4               0.0           0.231               0.0  
5               0.0           0.036               0.0  
6               0.0           0.057               0.0  
7               0.0           0.020               0.0  
8               0.0           0.013               1.0  
9               0.0           0.030               0.0  

[10 rows x 23 columns]

In [101]:
df8.isnull().sum()

participant_id          0
urine_barium          295
barium_comment        295
urine_cadmium         295
cadmium_comment       295
urine_cobalt          296
cobalt_comment        296
urine_cesium          295
cesium_comment        295
urine_molybdenum      295
molybdenum_comment    295
urine_manganese       295
manganese_comment     295
urine_lead            953
lead_comment          953
urine_antimony        295
antimony_comment      295
urine_tin             295
tin_comment           295
urine_thallium        295
thallium_comment      295
urine_tungsten        295
tungsten_comment      295
dtype: int64

As done prior, the plan is to delete rows that are missing most, if not all, of these values. 

In [102]:
df8.columns.to_list()

['participant_id',
 'urine_barium',
 'barium_comment',
 'urine_cadmium',
 'cadmium_comment',
 'urine_cobalt',
 'cobalt_comment',
 'urine_cesium',
 'cesium_comment',
 'urine_molybdenum',
 'molybdenum_comment',
 'urine_manganese',
 'manganese_comment',
 'urine_lead',
 'lead_comment',
 'urine_antimony',
 'antimony_comment',
 'urine_tin',
 'tin_comment',
 'urine_thallium',
 'thallium_comment',
 'urine_tungsten',
 'tungsten_comment']

In [103]:
metal_value_cols = [
    'urine_barium',
 'barium_comment',
 'urine_cadmium',
 'cadmium_comment',
 'urine_cobalt',
 'cobalt_comment',
 'urine_cesium',
 'cesium_comment',
 'urine_molybdenum',
 'molybdenum_comment',
 'urine_manganese',
 'manganese_comment',
 'urine_lead',
 'lead_comment',
 'urine_antimony',
 'antimony_comment',
 'urine_tin',
 'tin_comment',
 'urine_thallium',
 'thallium_comment',
 'urine_tungsten',
 'tungsten_comment']

rows_all_nan = df8[metal_value_cols].isna().all(axis=1)
rows_missing_all_metal = df8[rows_all_nan]

print(f"Number of rows missing all metal values: {rows_all_nan.sum()}")

Number of rows missing all metal values: 295


In [104]:
df8_clean = df8.dropna(how='all', subset=metal_value_cols)

In [105]:
df8_clean.isnull().sum()

participant_id          0
urine_barium            0
barium_comment          0
urine_cadmium           0
cadmium_comment         0
urine_cobalt            1
cobalt_comment          1
urine_cesium            0
cesium_comment          0
urine_molybdenum        0
molybdenum_comment      0
urine_manganese         0
manganese_comment       0
urine_lead            658
lead_comment          658
urine_antimony          0
antimony_comment        0
urine_tin               0
tin_comment             0
urine_thallium          0
thallium_comment        0
urine_tungsten          0
tungsten_comment        0
dtype: int64

##### Nickel

Nickel is another heavy metal that can potentially cause health risks if exposed at high levels. The most common reaction to nickel is contact dermatitis; however, inhalation and ingestion is also a possibility. When assessing urine samples, greater than 10 mg/dL in nickel concentration may indicate excessive exposure and calls for thorough evaluation [11].

In [106]:
file_path = '2017-2020/urine/10.P_UNI.xpt'

df9, meta = pyreadstat.read_xport(file_path)

In [107]:
df9.columns.to_list()

['SEQN', 'WTSAPRP', 'URXUNI', 'URDUNILC']

In [108]:
df9 = df9.drop('WTSAPRP', axis=1)

In [109]:
df9 = df9.rename(columns={
    'SEQN':'participant_id',
    'URXUNI': 'urine_nickel',
    'URDUNILC': 'urine_nickel_comment'
})

In [110]:
df9 = df9.dropna(subset=['urine_nickel'])

In [111]:
df9.head(10)

participant_id  urine_nickel  urine_nickel_comment
0        109266.0          0.46                   0.0
1        109270.0          1.08                   0.0
2        109273.0          0.91                   0.0
3        109274.0          1.17                   0.0
4        109287.0          4.01                   0.0
5        109288.0          1.08                   0.0
6        109290.0          2.72                   0.0
7        109295.0          0.22                   1.0
8        109300.0          0.55                   0.0
9        109309.0          0.92                   0.0

##### Organophosphate Insecticides

The lower limit of detection for the insecticide is 0.1 ng/mL.

In [112]:
file_path = '2017-2020/urine/11.P_OPD.xpt'

df10, meta = pyreadstat.read_xport(file_path)

In [113]:
df10.head(10)

SEQN       WTSBPRP  URXOP1  URDOP1LC  URXOP2  URDOP2LC  URXOP3  \
0  109271.0  20156.439742   0.682       0.0   0.288       0.0  0.1790   
1  109277.0  51738.369518   4.450       0.0  15.400       0.0  2.5900   
2  109282.0  97190.554500   1.580       0.0  21.000       0.0  1.3100   
3  109285.0  85548.221421   1.260       0.0   1.300       0.0  0.4900   
4  109288.0   9103.868995   0.196       0.0   1.120       0.0  0.0707   
5  109301.0  14215.937370   1.580       0.0   1.430       0.0  0.5190   
6  109302.0   5984.497323   3.290       0.0  35.800       0.0  1.1800   
7  109303.0  16549.099643   0.426       0.0     NaN       NaN  0.2020   
8  109304.0  40089.354988   1.520       0.0   3.530       0.0  1.5000   
9  109307.0  49745.101247   4.050       0.0   0.825       0.0     NaN   

   URDOP3LC  URXOP4  URDOP4LC  URXOP5  URDOP5LC  URXOP6  URDOP6LC  
0       0.0  0.0707       1.0  0.0707       1.0  0.0707       1.0  
1       0.0  0.2810       0.0  0.3770       0.0  0.0707       1.0  
2       0.0  1.0800       0.0  0.2600       0.0  0.0707       1.0  
3       0.0  0.0707       1.0  0.0707       1.0  0.0707       1.0  
4       1.0  0.0707       1.0  0.0707       1.0  0.0707       1.0  
5       0.0  0.0707       1.0  0.1500       0.0  0.0707       1.0  
6       0.0  1.1300       0.0  0.4560       0.0  0.0707       1.0  
7       0.0  0.0707       1.0  0.0707       1.0  0.0707       1.0  
8       0.0  0.0707       1.0  0.1040       0.0  0.0707       1.0  
9       NaN  0.1270       0.0  0.0707       1.0  0.0707       1.0

In [114]:
df10.shape

(4929, 14)

In [115]:
df10.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4929 entries, 0 to 4928
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   SEQN      4929 non-null   float64
 1   WTSBPRP   4929 non-null   float64
 2   URXOP1    4618 non-null   float64
 3   URDOP1LC  4618 non-null   float64
 4   URXOP2    4607 non-null   float64
 5   URDOP2LC  4607 non-null   float64
 6   URXOP3    4604 non-null   float64
 7   URDOP3LC  4604 non-null   float64
 8   URXOP4    4611 non-null   float64
 9   URDOP4LC  4611 non-null   float64
 10  URXOP5    4621 non-null   float64
 11  URDOP5LC  4621 non-null   float64
 12  URXOP6    4620 non-null   float64
 13  URDOP6LC  4620 non-null   float64
dtypes: float64(14)
memory usage: 539.2 KB


In [116]:
df10.isnull().sum()

SEQN          0
WTSBPRP       0
URXOP1      311
URDOP1LC    311
URXOP2      322
URDOP2LC    322
URXOP3      325
URDOP3LC    325
URXOP4      318
URDOP4LC    318
URXOP5      308
URDOP5LC    308
URXOP6      309
URDOP6LC    309
dtype: int64

In [117]:
df10.columns.to_list()

['SEQN',
 'WTSBPRP',
 'URXOP1',
 'URDOP1LC',
 'URXOP2',
 'URDOP2LC',
 'URXOP3',
 'URDOP3LC',
 'URXOP4',
 'URDOP4LC',
 'URXOP5',
 'URDOP5LC',
 'URXOP6',
 'URDOP6LC']

In [118]:
df10 = df10.drop('WTSBPRP', axis = 1)

In [119]:
df10 = df10.rename(columns={
    'SEQN' :'participant_id', 
 'URXOP1': 'dimethylphosphate_ng_mL',
 'URDOP1LC' : 'dimethylphosphate_comment',
 'URXOP2' : 'diethylphosphate_ng_mL',
 'URDOP2LC' : 'diethylphosphate_comment',
 'URXOP3' : 'dimethylthiophosphate_ng_mL',
 'URDOP3LC' :'dimethylthiophosphate_comment',
 'URXOP4': 'diethylthiophosphate_ng_mL',
 'URDOP4LC' :'diethylthiophosphate_comment',
 'URXOP5':'dimethyldithiophosphate_ng_mL',
 'URDOP5LC' :'dimethyldithiophosphate_comment' ,
 'URXOP6':'diethyldithiophosphate_ng_mL',
 'URDOP6LC':'diethyldithiophosphate_comment'
})

In [120]:
df10.columns.to_list()

['participant_id',
 'dimethylphosphate_ng_mL',
 'dimethylphosphate_comment',
 'diethylphosphate_ng_mL',
 'diethylphosphate_comment',
 'dimethylthiophosphate_ng_mL',
 'dimethylthiophosphate_comment',
 'diethylthiophosphate_ng_mL',
 'diethylthiophosphate_comment',
 'dimethyldithiophosphate_ng_mL',
 'dimethyldithiophosphate_comment',
 'diethyldithiophosphate_ng_mL',
 'diethyldithiophosphate_comment']

In [121]:
opd_value_cols = [
    'dimethylphosphate_ng_mL',
 'dimethylphosphate_comment',
 'diethylphosphate_ng_mL',
 'diethylphosphate_comment',
 'dimethylthiophosphate_ng_mL',
 'dimethylthiophosphate_comment',
 'diethylthiophosphate_ng_mL',
 'diethylthiophosphate_comment',
 'dimethyldithiophosphate_ng_mL',
 'dimethyldithiophosphate_comment',
 'diethyldithiophosphate_ng_mL',
 'diethyldithiophosphate_comment']

rows_all_nan = df10[opd_value_cols].isna().all(axis=1)
rows_missing_all_fr = df10[rows_all_nan]

print(f"Number of rows missing all OPD values: {rows_all_nan.sum()}")

Number of rows missing all OPD values: 307


In [122]:
df10_clean = df10.dropna(how='all', subset=opd_value_cols)

##### Perchlorate, Nitrate & Thiocyanate



In [123]:
file_path = '2017-2020/urine/12.P_PERNT.xpt'

df11, meta = pyreadstat.read_xport(file_path)

In [124]:
df11.head(10)

SEQN       WTSAPRP  URXUP8  URDUP8LC   URXNO3  URDNO3LC  URXSCN  \
0  109266.0  28660.015986   0.570       0.0  36900.0       0.0   223.0   
1  109270.0  17900.682903   4.020       0.0  48200.0       0.0  2960.0   
2  109273.0  80106.859617   2.170       0.0  49900.0       0.0  4740.0   
3  109274.0  24512.276280   6.950       0.0   2410.0       0.0  5290.0   
4  109287.0  25828.523003   7.290       0.0  78700.0       0.0   603.0   
5  109288.0   8535.018174   2.140       0.0  50600.0       0.0   370.0   
6  109290.0  12410.268374   2.970       0.0  70600.0       0.0  1470.0   
7  109295.0  28235.246814   2.140       0.0  24700.0       0.0  1010.0   
8  109300.0  66737.887353   0.876       0.0  10800.0       0.0   270.0   
9  109309.0  33019.729726   1.510       0.0  16800.0       0.0   171.0   

   URDSCNLC  
0       0.0  
1       0.0  
2       0.0  
3       0.0  
4       0.0  
5       0.0  
6       0.0  
7       0.0  
8       0.0  
9       0.0

In [125]:
df11.columns.to_list()

['SEQN',
 'WTSAPRP',
 'URXUP8',
 'URDUP8LC',
 'URXNO3',
 'URDNO3LC',
 'URXSCN',
 'URDSCNLC']

In [126]:
df11 = df11.drop('WTSAPRP', axis =1)

In [127]:
df11 = df11.rename(columns = {
    'SEQN': 'participant_id',
 'URXUP8': 'perchlorate_urine_ng_mL',
 'URDUP8LC': 'perchlorate_comment',
 'URXNO3':'nitrate_urine_ng_mL',
 'URDNO3LC':'nitrate_comment',
 'URXSCN':'thiocyanate_urine_ng_mL',
 'URDSCNLC': 'thiocyanate_comment'
})

In [128]:
df11.columns.to_list()

['participant_id',
 'perchlorate_urine_ng_mL',
 'perchlorate_comment',
 'nitrate_urine_ng_mL',
 'nitrate_comment',
 'thiocyanate_urine_ng_mL',
 'thiocyanate_comment']

In [129]:
pernt_value_cols = [
  'perchlorate_urine_ng_mL',
 'perchlorate_comment',
 'nitrate_urine_ng_mL',
 'nitrate_comment',
 'thiocyanate_urine_ng_mL',
 'thiocyanate_comment']

rows_all_nan = df11[pernt_value_cols].isna().all(axis=1)
rows_missing_all_fr = df11[rows_all_nan]

print(f"Number of rows missing all PERNT values: {rows_all_nan.sum()}")

Number of rows missing all PERNT values: 391


In [130]:
df11_clean = df11.dropna(how='all', subset=pernt_value_cols)

##### Urine Pregnancy Test

Point of care urine pregnancy test was performed on women 20-44 years of age. 

In [131]:
file_path = '2017-2020/urine/13.P_UCPREG.xpt'

df12, meta = pyreadstat.read_xport(file_path)

In [132]:
df12.head()

SEQN  URXPREG
0  109266.0      2.0
1  109284.0      2.0
2  109286.0      1.0
3  109291.0      2.0
4  109297.0      2.0

In [133]:
df12 = df12.rename(columns = {
    'SEQN': 'participant_id',
    'URXPREG' : 'pregnancy_test_result'
})

In [134]:
df12 = df12.dropna(subset=['pregnancy_test_result'])

##### Volatile Organic Compound (VOC) Metabolites

On the NHANES lab dataset, there are two VOC tables: P_UVOC and P_UVOC2. 


In [135]:
file_path = '2017-2020/urine/14.P_UVOC.xpt'

df13, meta = pyreadstat.read_xport(file_path)

In [136]:
df13.columns.to_list()

['SEQN',
 'WTSAPRP',
 'URX2MH',
 'URD2MHLC',
 'URX34M',
 'URD34MLC',
 'URXAAM',
 'URDAAMLC',
 'URXAMC',
 'URDAMCLC',
 'URXATC',
 'URDATCLC',
 'URXBMA',
 'URDBMALC',
 'URXBPM',
 'URDBPMLC',
 'URXCEM',
 'URDCEMLC',
 'URXCYHA',
 'URDCYALC',
 'URXCYM',
 'URDCYMLC',
 'URXDHB',
 'URDDHBLC',
 'URXGAM',
 'URDGAMLC',
 'URXHEM',
 'URDHEMLC',
 'URXHP2',
 'URDHP2LC',
 'URXHPM',
 'URDHPMLC',
 'URXIPM3',
 'URDPM3LC',
 'URXMAD',
 'URDMADLC',
 'URXMB3',
 'URDMB3LC',
 'URXPHG',
 'URDPHGLC',
 'URXPMM',
 'URDPMMLC',
 'URXTTC',
 'URDTTCLC']

In [137]:
import requests
from bs4 import BeautifulSoup
import pandas as pd

url = "https://wwwn.cdc.gov/Nchs/Data/Nhanes/Public/2017/DataFiles/P_UVOC.htm"
response = requests.get(url)
soup = BeautifulSoup(response.content, 'html.parser')

# Get the first table
table = soup.find('table')
rows = table.find_all('tr')

# Extract header and rows
headers = [th.get_text(strip=True) for th in rows[0].find_all('th')]
data = [
    [td.get_text(strip=True) for td in row.find_all('td')]
    for row in rows[1:]
]

df_info = pd.DataFrame(data, columns=headers)

In [138]:
import re

# Clean text to snake_case
def to_snake_case(text):
    text = text.lower()
    text = re.sub(r"[^\w\s]", "", text)  # Remove punctuation
    text = re.sub(r"\s+", "_", text)     # Replace whitespace with underscores
    return text

# Step 1: Create initial rename_dict from table (matching only columns in df13)
rename_dict = {
    row["VARIABLE NAME"]: to_snake_case(row["ANALYTE NAME"])
    for _, row in df_info.iterrows()
    if row["VARIABLE NAME"] in df13.columns
}

# Step 2: Clean comment code columns (ending in "LC")
unit_suffixes = ["_ng_ml", "_ug_l", "_mg_dl", "_umol_l", "_nmol_l"]
comment_renames = {}

for col in df13.columns:
    if col.endswith("LC") and col not in rename_dict:
        base_col = col[:-2]  # Remove 'LC'
        match_col = base_col.replace("URD", "URX")

        if match_col in rename_dict:
            clean_name = rename_dict[match_col]

            # Strip any unit suffix
            for unit in unit_suffixes:
                if clean_name.endswith(unit):
                    clean_name = clean_name[: -len(unit)]
                    break

            comment_renames[col] = f"{clean_name}_comment"

# Step 3: Merge comment renames into main rename_dict
rename_dict.update(comment_renames)

In [139]:
df13_cleaned = df13.rename(columns=rename_dict)

In [140]:
df13_cleaned.head()

SEQN       WTSAPRP  2methylhippuric_acid_ngml  \
0  109266.0  28660.015986                       3.54   
1  109270.0  17900.682903                       9.33   
2  109273.0  80106.859617                      90.10   
3  109274.0  24512.276280                      39.90   
4  109287.0  25828.523003                      14.70   

   2methylhippuric_acid_ngml_comment  3_and_4methylhippuric_acid_ngml  \
0                                1.0                             18.7   
1                                0.0                            106.0   
2                                0.0                            491.0   
3                                0.0                            118.0   
4                                0.0                            121.0   

   3_and_4methylhippuric_acid_ngml_comment  \
0                                      0.0   
1                                      0.0   
2                                      0.0   
3                                      0.0   
4                                      0.0   

   nacetyls2carbamoylethyllcysteine_ngml  \
0                                   19.0   
1                                  184.0   
2                                  122.0   
3                                   90.0   
4                                  135.0   

   nacetyls2carbamoylethyllcysteine_ngml_comment  \
0                                            0.0   
1                                            0.0   
2                                            0.0   
3                                            0.0   
4                                            0.0   

   nacetylsnmethylcarbamoyllcysteine_ngml  \
0                                    12.9   
1                                    91.0   
2                                   543.0   
3                                   129.0   
4                                    64.0   

   nacetylsnmethylcarbamoyllcysteine_ngml_comment  ...  mandelic_acid_ngml  \
0                                             0.0  ...                37.2   
1                                             0.0  ...               210.0   
2                                             0.0  ...               213.0   
3                                             0.0  ...               222.0   
4                                             0.0  ...              1270.0   

   mandelic_acid_ngml_comment  nacetyls4hydroxy2butenyllcysteine_ngml  \
0                         0.0                                   0.424   
1                         0.0                                   5.630   
2                         0.0                                  31.200   
3                         0.0                                   9.710   
4                         0.0                                  11.200   

   nacetyls4hydroxy2butenyllcysteine_ngml_comment  phenylglyoxylic_acid_ngml  \
0                                             1.0                       75.6   
1                                             0.0                      298.0   
2                                             0.0                      438.0   
3                                             0.0                      230.0   
4                                             0.0                      807.0   

   phenylglyoxylic_acid_ngml_comment  \
0                                0.0   
1                                0.0   
2                                0.0   
3                                0.0   
4                                0.0   

   nacetyls3hydroxypropyl1methyllcysteine_ngml  \
0                                         60.6   
1                                        257.0   
2                                        927.0   
3                                       2350.0   
4                                        240.0   

   nacetyls3hydroxypropyl1methyllcysteine_ngml_comment  \
0                                                0.0     
1                                                0.0     


In [141]:
df13_cleaned = df13_cleaned.drop('WTSAPRP',axis=1)

In [142]:
df13_cleaned.isnull().sum()

SEQN                                                       0
2methylhippuric_acid_ngml                                565
2methylhippuric_acid_ngml_comment                        565
3_and_4methylhippuric_acid_ngml                          565
3_and_4methylhippuric_acid_ngml_comment                  565
nacetyls2carbamoylethyllcysteine_ngml                    565
nacetyls2carbamoylethyllcysteine_ngml_comment            565
nacetylsnmethylcarbamoyllcysteine_ngml                   565
nacetylsnmethylcarbamoyllcysteine_ngml_comment           565
2aminothiazoline4carboxylic_acid_ngml                    565
2aminothiazoline4carboxylic_acid_ngml_comment            565
nacetylsbenzyllcysteine_ngml                             565
nacetylsbenzyllcysteine_ngml_comment                     565
nacetylsnpropyllcysteine_ngml                            565
nacetylsnpropyllcysteine_ngml_comment                    565
nacetyls2carboxyethyllcysteine_ngml                      565
nacetyls2carboxyethyllcy

In [143]:
# The three columns were missed since they did not follow the exact pattern. They were manually renamed

df13_cleaned = df13_cleaned.rename(columns ={
    'SEQN':'participant_id',
    'URDCYALC':'n_acetyl_s_1_cyano_2_hydroxyethyl_l_cysteine_comment',
    'URDPM3LC': 'n_acetyl_s_4_hydroxy_2_methyl_2_butenyl_l_cysteine_comment'
})

In [144]:
df13_cleaned.columns.to_list()

['participant_id',
 '2methylhippuric_acid_ngml',
 '2methylhippuric_acid_ngml_comment',
 '3_and_4methylhippuric_acid_ngml',
 '3_and_4methylhippuric_acid_ngml_comment',
 'nacetyls2carbamoylethyllcysteine_ngml',
 'nacetyls2carbamoylethyllcysteine_ngml_comment',
 'nacetylsnmethylcarbamoyllcysteine_ngml',
 'nacetylsnmethylcarbamoyllcysteine_ngml_comment',
 '2aminothiazoline4carboxylic_acid_ngml',
 '2aminothiazoline4carboxylic_acid_ngml_comment',
 'nacetylsbenzyllcysteine_ngml',
 'nacetylsbenzyllcysteine_ngml_comment',
 'nacetylsnpropyllcysteine_ngml',
 'nacetylsnpropyllcysteine_ngml_comment',
 'nacetyls2carboxyethyllcysteine_ngml',
 'nacetyls2carboxyethyllcysteine_ngml_comment',
 'nacetyls1cyano2hydroxyethyllcysteine_ngml',
 'n_acetyl_s_1_cyano_2_hydroxyethyl_l_cysteine_comment',
 'nacetyls2cyanoethyllcysteine_ngml',
 'nacetyls2cyanoethyllcysteine_ngml_comment',
 'nacetyls34dihydroxybutyllcysteine_ngml',
 'nacetyls34dihydroxybutyllcysteine_ngml_comment',
 'nacetyls2carbamoyl2hydroxyethyllcy

In [145]:
file_path = '2017-2020/urine/15.P_UVOC2.xpt'

df14, meta = pyreadstat.read_xport(file_path)

In [146]:
df14.head()

SEQN      WTVOC2PP  URXMUCA  URDMUCLC  URXPHMA  URDPMALC
0  109266.0  29122.785906     6.94       1.0    0.106       1.0
1  109270.0  18436.336755   213.00       0.0    0.185       0.0
2  109273.0  93177.905637   147.00       0.0    1.100       0.0
3  109274.0  27374.984127   244.00       0.0    0.171       0.0
4  109287.0  25946.229537   130.00       0.0    0.153       0.0

In [147]:
df14 = df14.rename(columns={
    'SEQN': 'participant_id',
    'URXMUCA': 'trans_trans_muconic_acid_ng_mL',
    'URDMUCLC': 'trans_trans_muconic_acid_comment',
    'URXPHMA': 'phenylmercapturic_acid_ng_mL',
    'URDPMALC': 'phenylmercapturic_acid_comment'
})   

In [148]:
df14 = df14.drop('WTVOC2PP', axis=1)

In [149]:
df14.isnull().sum()

participant_id                        0
trans_trans_muconic_acid_ng_mL      994
trans_trans_muconic_acid_comment    994
phenylmercapturic_acid_ng_mL        994
phenylmercapturic_acid_comment      994
dtype: int64

In [150]:
common_nan = get_common_nan_ids(df14, 'trans_trans_muconic_acid_ng_mL', 'phenylmercapturic_acid_ng_mL', id_col='participant_id')

Number of NaNs in trans_trans_muconic_acid_ng_mL: 994
Number of NaNs in phenylmercapturic_acid_ng_mL: 994
Number of IDs with NaNs in both columns: 994


In [151]:
df14_cleaned = drop_rows_with_common_nan_ids(df14, 'trans_trans_muconic_acid_ng_mL', 'phenylmercapturic_acid_ng_mL', id_col='participant_id')

Rows dropped where both trans_trans_muconic_acid_ng_mL and phenylmercapturic_acid_ng_mL were NaN: 994


All of the urine labs have been cleaned. The dataframes from all of the urine labs will be collated into one large dataframe named urine_labs into csv file.

In [152]:
df_names = [var for var in globals() if isinstance(globals()[var], pd.DataFrame)]
print(df_names)

['__', 'df_demo', '_3', '_8', '_12', '_14', '_17', 'df_bmx', '_20', '_28', '_30', 'df', '_35', 'df_cleaned', 'df1', '_44', 'df2', '_55', 'df2_cleaned', 'df3', '_61', 'df4', '_68', 'rows_missing_all_fr', 'df4_clean', '_75', 'df5', 'df5_cleaned', 'df6', 'df6_clean', 'df7', '_90', 'df7_clean', '_95', 'df8', '_100', 'rows_missing_all_metal', 'df8_clean', 'df9', '_111', 'df10', '_113', 'df10_clean', 'df11', '_124', 'df11_clean', 'df12', '_132', 'df13', 'df_info', 'df13_cleaned', '_140', 'df14', '_146', 'df14_cleaned']


In [153]:
urine_dfs = [
    df_cleaned,
    df1,
    df2_cleaned,
    df3,
    df4_clean,
    df5_cleaned,
    df6_clean,
    df7_clean,
    df8_clean,
    df9,
    df10_clean,
    df11_clean,
    df12,
    df13_cleaned,
    df14_cleaned
]

from functools import reduce

df_urine_combined = reduce(
    lambda left, right: pd.merge(left, right, on="participant_id", how="outer"),
    urine_dfs
)

In [154]:
df_urine_combined.to_csv("cleaned_urine_labs_combined.csv", index=False)

In [161]:
urine_df = pd.read_csv('cleaned_urine_labs_combined.csv')

urine_df.head(10)

participant_id  albumin_urine_ug_mL  albumin_urine_mg_L  alb_comment  \
0        109266.0                  5.5                 5.5          0.0   
1        109270.0                  4.0                 4.0          0.0   
2        109271.0                  2.4                 2.4          0.0   
3        109273.0                  4.9                 4.9          0.0   
4        109274.0                 12.8                12.8          0.0   
5        109275.0                  3.7                 3.7          0.0   
6        109277.0                 14.7                14.7          0.0   
7        109278.0                  8.4                 8.4          0.0   
8        109279.0                 13.9                13.9          0.0   
9        109282.0                 16.0                16.0          0.0   

   creatinine_urine_mg_dL  creatinine_urine_umol_L  creatinine_comment  \
0                    36.0                   3182.4                 0.0   
1                   165.0                  14586.0                 0.0   
2                    32.0                   2828.8                 0.0   
3                   121.0                  10696.4                 0.0   
4                   120.0                  10608.0                 0.0   
5                    20.0                   1768.0                 0.0   
6                   244.0                  21569.6                 0.0   
7                   124.0                  10961.6                 0.0   
8                   251.0                  22188.4                 0.0   
9                   192.0                  16972.8                 0.0   

   alb_creat_ratio  arsenous_acid_ug_L  arsenous_acid_comment  ...  \
0            15.28                0.08                    1.0  ...   
1             2.42                0.51                    0.0  ...   
2             7.50                 NaN                    NaN  ...   
3             4.05                0.08                    1.0  ...   
4            10.67                0.08                    1.0  ...   
5            18.50                 NaN                    NaN  ...   
6             6.02                 NaN                    NaN  ...   
7             6.77                 NaN                    NaN  ...   
8             5.54                 NaN                    NaN  ...   
9             8.33                 NaN                    NaN  ...   

   phenylglyoxylic_acid_ngml  phenylglyoxylic_acid_ngml_comment  \
0                       75.6                                0.0   
1                      298.0                                0.0   
2                        NaN                                NaN   
3                      438.0                                0.0   
4                      230.0                                0.0   
5                        NaN                                NaN   
6                        NaN                                NaN   
7                        NaN                                NaN   
8                        NaN                                NaN   
9                        NaN                                NaN   

   nacetyls3hydroxypropyl1methyllcysteine_ngml  \
0                                         60.6   
1                                        257.0   
2                                          NaN   
3                                        927.0   
4                                       2350.0   
5                                          NaN   
6                                          NaN   
7                                          NaN   
8                                          NaN   
9                                          NaN   

   nacetyls3hydroxypropyl1methyllcysteine_ngml_comment  \
0                                                0.0     
1                                                0.0     
2                                                NaN     
3                                                0.0     
4           

In [162]:
urine_df.shape

(12795, 129)

#### Blood Labs

**Alpha-1-acid glycoprotein (AGP)**

Alpha-1-acid glycoprotein (AGP), also known as orosomucoid (ORM), is an acute-phase serum protein present in humans and many animal species. It is produced in response to inflammation, although its precise biological role remains under investigation and somewhat ambiguous [2]. According to Ceciliani et al. (2019), AGP may play a role in immunometabolism, a function potentially relevant to understanding the obesity epidemic in the U.S.

In the NHANES dataset, AGP levels were measured in children aged 3–5 years and females aged 12–49 years. This data offers an opportunity to explore potential correlations between AGP serum concentrations and obesity prevalence among the female participants in the study.

In [172]:
file_path = '2017-2020/blood/1.P_SSAGP.xpt'

df_b1, meta = pyreadstat.read_xport(file_path)

In [173]:
df_b1.head(10)

SEQN      WTSSAGPP  SSAGP
0  109264.0      0.000000    NaN
1  109266.0  10003.783188  0.796
2  109277.0  23329.384783  0.746
3  109279.0  14416.168293  0.940
4  109284.0  17705.030492  1.080
5  109286.0  21951.734438  0.358
6  109288.0      0.000000    NaN
7  109291.0  25981.105960  0.766
8  109297.0      0.000000    NaN
9  109309.0      0.000000    NaN

In [174]:
df_b1.shape

(3823, 3)

In [175]:
df_b1 = df_b1.rename(columns={
    'SEQN':'participant_id',
    'SSAGP': 'alpha_1_agp_g_L'
})

In [176]:
df_b1.head()

participant_id      WTSSAGPP  alpha_1_agp_g_L
0        109264.0      0.000000              NaN
1        109266.0  10003.783188            0.796
2        109277.0  23329.384783            0.746
3        109279.0  14416.168293            0.940
4        109284.0  17705.030492            1.080

In [177]:
df_b1 = df_b1.drop('WTSSAGPP', axis=1)

In [178]:
df_b1 = df_b1.dropna(subset = ['alpha_1_agp_g_L'])

In [179]:
df_b1.head(5)

participant_id  alpha_1_agp_g_L
1        109266.0            0.796
2        109277.0            0.746
3        109279.0            0.940
4        109284.0            1.080
5        109286.0            0.358

**Lipid Panel**

Lipids are essential molecules that support a range of physiological functions, including hormone production and cellular structure. However, excessive lipid levels—particularly certain types—are associated with increased risk of cardiovascular disease.

To assess lipid status, a fasting lipid panel is commonly used. This test typically includes measurements of:
- LDL (low-density lipoprotein, or “bad” cholesterol),
- HDL (high-density lipoprotein, or “good” cholesterol),
- Total cholesterol, and
- Triglycerides

The NHANES dataset includes all of these values, enabling analysis of lipid profiles across a large representative population. This section focuses on cleaning and preparing these variables for analysis.

In [180]:
file_path = '2017-2020/blood/2.P_HDL.xpt'

df_b2, meta = pyreadstat.read_xport(file_path)

In [181]:
df_b2.head()

SEQN  LBDHDD  LBDHDDSI
0  109264.0    72.0      1.86
1  109266.0    56.0      1.45
2  109270.0    47.0      1.22
3  109271.0    33.0      0.85
4  109273.0    42.0      1.09

## References

2. Ceciliani, F., & Lecchi, C. (2019). The immune functions of α1 acid glycoprotein. Current Protein & Peptide Science, 20(6), 505–524. https://doi.org/10.2174/1389203720666190405101138
3. National Institute of Environmental Health Sciences. (n.d.). Arsenic. U.S. Department of Health and Human Services. Retrieved July 3, 2025, from https://www.niehs.nih.gov/health/topics/agents/arsenic
4. Office of Dietary Supplements. (2022, June 2). Chromium: Health professional fact sheet. U.S. Department of Health and Human Services. Retrieved July 6, 2025, from https://ods.od.nih.gov/factsheets/Chromium-HealthProfessional/
5. National Institute of Environmental Health Sciences. (2025, January 30). Flame retardants. U.S. Department of Health and Human Services. Retrieved July 7, 2025, from https://www.niehs.nih.gov/health/topics/agents/flame_retardants
6. Office of Dietary Supplements. (2024, November 5). Iodine: Health professional fact sheet. U.S. Department of Health and Human Services, National Institutes of Health. Retrieved July 8, 2025, from https://ods.od.nih.gov/factsheets/Iodine-HealthProfessional/
7. Pearce, E. N., & Caldwell, K. L. (2016). Urinary iodine, thyroid function, and thyroglobulin as biomarkers of iodine status. The American Journal of Clinical Nutrition, 104(Suppl 3), 898S–901S. https://doi.org/10.3945/ajcn.115.110395
8. Agency for Toxic Substances and Disease Registry. (2007, August). Toxicological profile for barium and barium compounds: Public health statement. U.S. Department of Health and Human Services. https://www.ncbi.nlm.nih.gov/books/NBK598787/
9. Ye, B. J., Kim, B. G., Jeon, M. J., Kim, S. Y., Kim, H. C., Jang, T. W., Chae, H. J., Choi, W. J., Ha, M. N., & Hong, Y. S. (2016). Evaluation of mercury exposure level, clinical diagnosis and treatment for mercury intoxication. Annals of occupational and environmental medicine, 28, 5. https://doi.org/10.1186/s40557-015-0086-8
10. Genchi, G., Sinicropi, M. S., Lauria, G., Carocci, A., & Catalano, A. (2020). The Effects of Cadmium Toxicity. International journal of environmental research and public health, 17(11), 3782. https://doi.org/10.3390/ijerph17113782
11. Gates, A., Jakubowski, J. A., & Regina, A. C. (2025). Nickel toxicology. In StatPearls. StatPearls Publishing. Retrieved May 20, 2023, from https://www.ncbi.nlm.nih.gov/books/NBK592400/

